shop/src/main/java/ro/msg/learning/shop/service/StockService.java

```diff
@@ -0,0 +1,137 @@
+        Product product = productRepository.findById(stockCreateDto.getProductId())
+                .orElseThrow(() -> new ResourceNotFoundException(PRODUCT_WITH_ID + stockCreateDto.getProductId() + NOT_FOUND));
+
+        Location location = locationRepository.findById(stockCreateDto.getLocationId())
+                .orElseThrow(() -> new ResourceNotFoundException(LOCATION_WITH_ID + stockCreateDto.getLocationId() + NOT_FOUND));
+
+        Stock stock = new Stock();
+        stock.setProduct(product);
+        stock.setLocation(location);
+        stock.setQuantity(stockCreateDto.getQuantity());
+
+        stock = stockRepository.save(stock);
+        return stockMapper.mapStockToStockGetDto(stock);
+    }
+
+    public StockGetDto updateStock(StockId stockId, StockCreateDto stockCreateDto) {
+        validateProductAndLocationExistence(stockCreateDto.getProductId(), stockCreateDto.getLocationId());
+
+        Stock existingStock = stockRepository.findById(stockId)
+                .orElseThrow(() -> new ResourceNotFoundException(STOCK_WITH_ID + stockId + NOT_FOUND));
+
+        Product product = productRepository.getById(stockCreateDto.getProductId());
+        Location location = locationRepository.getById(stockCreateDto.getLocationId());
+
+        stockMapper.mapStockCreateDtoToStock(stockCreateDto, product, location);
+        existingStock.setProduct(product);
+        existingStock.setLocation(location);
+        existingStock.setQuantity(stockCreateDto.getQuantity());
+
+        stockRepository.save(existingStock);
+        return stockMapper.mapStockToStockGetDto(existingStock);
+    }
+
+    public void deleteStock(StockId stockId) {
+        if (!stockRepository.existsById(stockId)) {
+            throw new ResourceNotFoundException(STOCK_WITH_ID + stockId + NOT_FOUND);
+        }
+        stockRepository.deleteById(stockId);
+    }
+
+    private void validateProductAndLocationExistence(UUID productId, UUID locationId) {
+        if (!productRepository.existsById(productId)) {
+            throw new ResourceNotFoundException(PRODUCT_WITH_ID + productId + NOT_FOUND);
+        }
+
+        if (!locationRepository.existsById(locationId)) {
+            throw new ResourceNotFoundException(LOCATION_WITH_ID + locationId + NOT_FOUND);
+        }
+    }
+
+    public boolean hasSufficientStock(Location location, List<ProductIdAndQuantityDto> productList) {
+        for (ProductIdAndQuantityDto dto : productList) {
+            Product product = productRepository.getProductById(dto.getProductId());
+            if (product == null) {
+                throw new ResourceNotFoundException(PRODUCT_NOT_FOUND_WITH_ID  + dto.getProductId());
+            }
+
+            Stock stock = stockRepository.findByLocationAndProduct(location, product);
+
+            if (stock == null || stock.getQuantity() < dto.getQuantity()) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Transactional
```

Comment:
❤️ whooooaaaa, new annotation who dis?
nice use of transactional. This is a powerful annotation and can be used in order to make sure that a transaction (an action which happens upon the database and modifies fields) does not backfire by only updateing part of the fields and so on. 
However, in order to use it corectly and actually make it shine it needs to be configured in a @Configuration file. Please take 10 minutes and read the following article and then abuse transactional: https://www.baeldung.com/transaction-configuration-with-jpa-and-spring

Reply:
Couldn't quite figure out @Transactional annotation. I removed it but I'm keeping it in mind for future projects.